writer.go

```diff
@@ -19,28 +17,45 @@ type Writer struct {
 	framer    Framer
 	formatter Formatter
 
+	mu   sync.RWMutex // guards conn
 	conn serverConn
 }
 
+// getConn provides access to the internal conn, protected by a mutex. The
+// conn is threadsafe, so it can be used while unlocked, but we want to avoid
+// race conditions on grabbing a reference to it.
+func (w *Writer) getConn() serverConn {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	return w.conn
+}
+
+// setConn updates the internal conn, protected by a mutex.
+func (w *Writer) setConn(c serverConn) {
+	w.mu.Lock()
+	w.conn = c
+	w.mu.Unlock()
```

Comment:
It will, but `defer` has a small performance hit that we can pretty easily avoid here. (That's literally the only reason for not using `defer` here, in addition to the fact that in this case `defer` isn't necessary: it's most helpful when a function has multiple `return`s that we'd have to unlock in front of.)